### PR TITLE
[Feature] Stream load supports compression

### LINKED ISSUE #44642: Stream load supports compression
- [x] https://github.com/StarRocks/starrocks/pull/44985
- [x] https://github.com/StarRocks/starrocks/pull/44987
- [x] https://github.com/StarRocks/starrocks/pull/43732

## Feature request

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by rickif**
## Why I'm doing:
Now, stream load doesn't support compressed data payload.
## What I'm doing:
This PR adds the support of the compressed data payload in JSON format. You only need to set the compression type in `compression` or `Content-Encoding` HTTP header.
```
curl --location-trusted -u root:  -v 'http://127.0.0.1:18030/api/db0/tbl_simple/_stream_load' \-X PUT  -H "expect:100-continue" \-H 'format: json' -H 'compression: lz4_frame'   -T ./b.json.lz4

curl --location-trusted -u root:  -v 'http://127.0.0.1:18030/api/db0/tbl_simple/_stream_load' \-X PUT  -H "expect:100-continue" \-H 'format: json' -H 'Content-Encoding: lz4_frame'   -T ./b.json.lz4
```
Compression supported:
1. gzip
2. bzip2
3. LZ4_FRAME
4. DEFLATE
5. ZSTD

Fixes https://github.com/StarRocks/starrocks/issues/44642

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 1]
```diff
@@ -3517,6 +3517,8 @@ public PlanFragment visitPhysicalTableFunctionTableScan(OptExpression optExpress
             prepareContextSlots(node, context, tupleDesc);
 
             int dop = ConnectContext.get().getSessionVariable().getSinkDegreeOfParallelism();
+            int dop2 = context.getConnectContext().getSessionVariable().getSinkDegreeOfParallelism();
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:3520]: seems unnecessary change introduced.

File: fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java [hunk 2]
```diff
@@ -450,6 +455,10 @@ private void setOptionalFromTSLPutRequest(TStreamLoadPutRequest request, Databas
         if (request.isSetPartial_update_mode()) {
             partialUpdateMode = request.getPartial_update_mode();
         }
+
+        if (request.isSetCompression_type()) {
+            payloadCompressionType = CompressionUtils.findTCompressionByName(request.getCompression_type());
```
[the remark this answers is not in the document]
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java]: > And, if the `compression` is not set, can SR get the compression method from the file extension?

An error "unsupported compression type: xxx" would be reported

[the remark this answers is not in the document]
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java]: > Will it be Ok if the specified compression method is not supported?

An error "unsupported compression type: xxx" would be reported

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 3]
```diff
@@ -275,6 +275,81 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
...
+        return ret.status();
+    }
+    auto buf = ret.value();
+
+    // try to read all compressed data into _decompressed_buffer
+    bool stream_end = false;
+    size_t total_bytes_read = 0;
+    size_t bytes_read = 0;
+    size_t bytes_written = 0;
+    RETURN_IF_ERROR(_decompressor->decompress(reinterpret_cast<uint8_t*>(buf->ptr), buf->remaining(), &bytes_read,
```
[the remark this answers is not in the document]
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.cpp:316]: If it's not enough, the `stream_end` will be false. Then we will try to read all payload in multi buffers.

**Comment by banmoy** [comment]
How to support compression for transaction stream load?

**Comment by rickif** [comment]
> How to support compression for transaction stream load?

The support of the header `Content-Encoding` for transaction stream load is added now.

File: be/src/runtime/stream_load/stream_load_pipe.h [hunk 4]
```diff
@@ -108,6 +110,19 @@ class StreamLoadPipe : public MessageBodySink {
...
 };
 
+const size_t DEFAULT_DECOMPRESS_BUFFER_SIZE = 32 * 1024 * 1024;
+class DecompressedStreamLoadPipe : public StreamLoadPipe {
+public:
+    DecompressedStreamLoadPipe(std::shared_ptr<StreamLoadPipe> pipe, TCompressionType::type compression_type);
+    ~DecompressedStreamLoadPipe() override = default;
+    StatusOr<ByteBufferPtr> read() override;
+private:
+    std::shared_ptr<StreamLoadPipe> _pipe;
```
[the remark this answers is not in the document]
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.h]: Already removed.

File: be/src/util/compression/stream_compression.cpp [hunk 5]
```diff
@@ -371,6 +372,19 @@ Status Lz4FrameStreamCompression::decompress(uint8_t* input, size_t input_len, s
     return Status::OK();
 }
 
+Status Lz4FrameStreamCompression::decompress_header(uint8_t* input, size_t input_len, size_t* decompressed_len) {
```
[the remark this answers is not in the document]
Comment (rickif) [reply be/src/util/compression/stream_compression.cpp]: Already removed.

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 6]
```diff
@@ -275,6 +275,81 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
     return Status::OK();
 }
 
+DecompressedStreamLoadPipe::DecompressedStreamLoadPipe(std::shared_ptr<StreamLoadPipe> pipe,
+                                                       TCompressionType::type compression_type)
+        : _pipe(pipe), _compression_type(compression_type) {}
+
+StatusOr<ByteBufferPtr> DecompressedStreamLoadPipe::read() {
+    if (_decompressor == nullptr) {
+        CompressionTypePB compression = CompressionTypePB::DEFAULT_COMPRESSION;
```
[the remark this answers is not in the document]
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.cpp]: Updated

File: be/src/http/http_common.h [hunk 7]
```diff
@@ -75,5 +75,6 @@ static const std::string HTTP_PARTIAL_UPDATE_MODE = "partial_update_mode";
 
 static const std::string HTTP_100_CONTINUE = "100-continue";
 static const std::string HTTP_CHANNEL_ID = "channel_id";
+static const std::string HTTP_COMPRESSION = "compression";
```
Comment (kevincai) [inline be/src/http/http_common.h:78]: not suggest go this way. The standard Content-Encoding is doing exactly the same thing, customized header should be used as few as possible.

File: be/src/runtime/stream_load/stream_load_pipe.h [hunk 8]
```diff
@@ -110,14 +110,24 @@ class StreamLoadPipe : public MessageBodySink {
...
+public:
+    StreamLoadPipeReader(std::shared_ptr<StreamLoadPipe> pipe) : _pipe(pipe) {}
+    virtual ~StreamLoadPipeReader() {}
+    virtual StatusOr<ByteBufferPtr> read() { return _pipe->read(); }
+
+private:
+    std::shared_ptr<StreamLoadPipe> _pipe;
+};
+
 const size_t DEFAULT_DECOMPRESS_BUFFER_SIZE = 32 * 1024 * 1024;
```
Comment (jaogoy) [inline be/src/runtime/stream_load/stream_load_pipe.h:123]: If DEFAULT_DECOMPRESS_BUFFER_SIZE is only used in CompressedStreamLoadPipeReader, put it in.

File: fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java [hunk 9]
```diff
@@ -457,7 +457,11 @@ private void setOptionalFromTSLPutRequest(TStreamLoadPutRequest request, Databas
         }
 
         if (request.isSetCompression_type()) {
-            payloadCompressionType = CompressionUtils.findTCompressionByName(request.getCompression_type());
+            TCompressionType compression = CompressionUtils.findTCompressionByName(request.getCompression_type());
+            if (compression == null) {
+                throw new UserException("unsupported compression type: " + request.getCompression_type());
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java:462]: Put this error message into ErrorCode.java if it will be shown to users.

File: fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java [hunk 10]
```diff
@@ -457,7 +457,11 @@ private void setOptionalFromTSLPutRequest(TStreamLoadPutRequest request, Databas
...
+            }
+            payloadCompressionType = compression;
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/load/streamload/StreamLoadInfo.java:464]: Why not directly use ` payloadCompressionType = CompressionUtils.xxx` rather than using a new variable `compression`?

File: be/src/exec/json_scanner.cpp [hunk 11]
```diff
@@ -683,10 +683,11 @@ Status JsonReader::_construct_row(simdjson::ondemand::object* row, Chunk* chunk)
 
 Status JsonReader::_read_file_stream() {
     // TODO: Remove the down_cast, should not rely on the specific implementation.
-    auto pipe = down_cast<StreamLoadPipeInputStream*>(_file->stream().get())->pipe();
+    auto pipe = make_shared<StreamLoadPipeReader>(down_cast<StreamLoadPipeInputStream*>(_file->stream().get())->pipe());
```
Comment (jaogoy) [inline be/src/exec/json_scanner.cpp:686]: 1. It's better to rename to pipeReader.
2. The StreamLoadPipeReader object will be created even if using compressedReader, right?

File: be/src/http/action/stream_load.cpp [hunk 12]
```diff
@@ -586,6 +586,10 @@ Status StreamLoadAction::_process_put(HttpRequest* http_req, StreamLoadContext*
         ctx->max_filter_ratio = strtod(http_req->header(HTTP_MAX_FILTER_RATIO).c_str(), nullptr);
     }
 
+    if (!http_req->header(HttpHeaders::CONTENT_ENCODING).empty()) {
+        request.__set_compression_type(http_req->header(HttpHeaders::CONTENT_ENCODING));
```
[the remark this answers is not in the document]
Comment (jaogoy) [reply be/src/http/action/stream_load.cpp]: @kevincai `compression` is a parameter for all loading methods.
It may be better to implement `content-encoding` besides `compression`.

File: be/src/http/http_common.h (hunk 7 above)
Comment (rickif) [reply be/src/http/http_common.h:78]: I agree. @jaogoy PTAL.

File: be/src/exec/json_scanner.cpp (hunk 11 above)
Comment (rickif) [reply be/src/exec/json_scanner.cpp:686]: > 2. The StreamLoadPipeReader object will be created even if using compressedReader, right?

Yes.

Comment (rickif) [reply be/src/exec/json_scanner.cpp:686]: > 2. The StreamLoadPipeReader object will be created even if using compressedReader, right?

Yes.

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by rickif** [comment]
@mergifyio rebase

**Review by jaogoy — DISMISSED** [review]

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 13]
```diff
@@ -275,6 +277,67 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
...
+        pieces_size += bytes_written;
+        total_bytes_read += bytes_read;
+        bytes_read = 0;
+        bytes_written = 0;
+    }
+
+    if (pieces_size > 0) {
+        if (_decompressed_buffer->remaining() < pieces_size) {
+            // align to 1024 bytes.
+            auto sz = std::ceil((_decompressed_buffer->capacity + pieces_size) * 1.0 / 1024) * 1024;
```
Comment (wyb) [inline be/src/runtime/stream_load/stream_load_pipe.cpp:330]:
```suggestion
            auto sz = std::ceil((_decompressed_buffer->pos + pieces_size) * 1.0 / 1024) * 1024;
```

**Review by wyb — DISMISSED** [review]

**Comment by rickif** [comment]
@mergifyio rebase

**Review by meegoo — DISMISSED** [review]

File: gensrc/thrift/FrontendService.thrift [hunk 14]
```diff
@@ -913,6 +913,7 @@ struct TStreamLoadPutRequest {
     // only valid when file type is CSV
     54: optional byte escape
     55: optional Types.TPartialUpdateMode partial_update_mode
+    56: optional string compression_type
```
Comment (trueeyu) [inline gensrc/thrift/FrontendService.thrift:916]: Why not use Types.TCompressionType?

Comment (rickif) [reply gensrc/thrift/FrontendService.thrift:916]: It's also OK to use `Types.TCompressionType` here. To use `CompressionUtils.findTCompressionByName()` to convert string to `Types.TCompressionType`, I use a simple string here.

Comment (trueeyu) [reply gensrc/thrift/FrontendService.thrift:916]: Why fe need know this?

Comment (trueeyu) [reply gensrc/thrift/FrontendService.thrift:916]: rename to `payload_compression_type` ?

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 15]
```diff
@@ -275,6 +277,67 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
...
 
+CompressedStreamLoadPipeReader::CompressedStreamLoadPipeReader(std::shared_ptr<StreamLoadPipe> pipe,
+                                                               TCompressionType::type compression_type)
+        : StreamLoadPipeReader(std::move(pipe)), _compression_type(compression_type) {}
+
+StatusOr<ByteBufferPtr> CompressedStreamLoadPipeReader::read() {
+    if (_decompressor == nullptr) {
+        auto compression = CompressionUtils::to_compression_pb(_compression_type);
+        if (compression == CompressionTypePB::UNKNOWN_COMPRESSION) {
+            return Status::NotSupported("Unsupported compression algorithm: " + std::to_string(_compression_type));
```
Comment (trueeyu) [inline be/src/runtime/stream_load/stream_load_pipe.cpp:288]: std::to_string(_compression_type) convert to the string "LZ4" or "4"?

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 16]
```diff
@@ -275,6 +277,67 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
...
+        total_bytes_read += bytes_read;
+        bytes_read = 0;
+        bytes_written = 0;
+    }
+
+    if (pieces_size > 0) {
+        if (_decompressed_buffer->remaining() < pieces_size) {
+            // align to 1024 bytes.
+            auto sz = std::ceil((_decompressed_buffer->pos + pieces_size) * 1.0 / 1024) * 1024;
+            _decompressed_buffer = ByteBuffer::reallocate(_decompressed_buffer, sz);
```
Comment (trueeyu) [inline be/src/runtime/stream_load/stream_load_pipe.cpp:331]: why _decompressed_buffer need resize?

File: be/src/runtime/stream_load/stream_load_pipe.cpp (hunk 15 above)
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.cpp:288]: LZ4

File: be/src/runtime/stream_load/stream_load_pipe.h [hunk 17]
```diff
@@ -108,6 +110,29 @@ class StreamLoadPipe : public MessageBodySink {
...
+};
+
+class CompressedStreamLoadPipeReader : public StreamLoadPipeReader {
+public:
+    CompressedStreamLoadPipeReader(std::shared_ptr<StreamLoadPipe> pipe, TCompressionType::type compression_type);
+    ~CompressedStreamLoadPipeReader() override = default;
+    StatusOr<ByteBufferPtr> read() override;
+
+private:
+    const size_t DEFAULT_DECOMPRESS_BUFFER_SIZE = 32 * 1024 * 1024;
```
Comment (trueeyu) [inline be/src/runtime/stream_load/stream_load_pipe.h:130]: For all stream load, allocate one buffer of 32M?

Comment (trueeyu) [reply be/src/runtime/stream_load/stream_load_pipe.h:130]: If StreamLoad concurrency is relatively high, there is a risk of excessive memory usage.？

**Review by wyb — DISMISSED** [review]

**Review by packy92 — DISMISSED** [review]

**Review by kevincai — DISMISSED** [review]

**Review by wangruin — DISMISSED** [review]

File: be/src/util/byte_buffer.h [hunk 18]
```diff
@@ -52,6 +52,13 @@ struct ByteBuffer {
         return ptr;
     }
 
+    static ByteBufferPtr reallocate(const ByteBufferPtr& old_ptr, size_t new_size) {
+        ByteBufferPtr ptr(new ByteBuffer(new_size));
```
Comment (trueeyu) [inline be/src/util/byte_buffer.h:56]: Reallocate:  only support alloc larger.
Flip: make smaller and free up excess memory

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 19]
```diff
@@ -275,6 +277,70 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
...
+        pieces_size += bytes_written;
+        total_bytes_read += bytes_read;
+        bytes_read = 0;
+        bytes_written = 0;
+    }
+
+    if (pieces_size > 0) {
+        if (_decompressed_buffer->remaining() < pieces_size) {
+            // align to 1024 bytes.
+            auto sz = std::ceil((_decompressed_buffer->pos + pieces_size) * 1.0 / 1024) * 1024;
```
Comment (trueeyu) [inline be/src/runtime/stream_load/stream_load_pipe.cpp:333]: Is it possible that sz is 0?

File: be/src/http/http_common.h (hunk 7 above)
Comment (trueeyu) [reply be/src/http/http_common.h:78]: compression or compression_type

File: be/src/runtime/stream_load/stream_load_pipe.h (hunk 17 above)
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.h]: Now, the  buffer size grows exponentially

File: be/src/runtime/stream_load/stream_load_pipe.cpp [hunk 20]
```diff
@@ -275,6 +277,67 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
...
+    ASSIGN_OR_RETURN(auto buf, StreamLoadPipeReader::read());
+
+    // try to read all compressed data into _decompressed_buffer
+    bool stream_end = false;
+    size_t total_bytes_read = 0;
+    size_t bytes_read = 0;
+    size_t bytes_written = 0;
+    RETURN_IF_ERROR(_decompressor->decompress(reinterpret_cast<uint8_t*>(buf->ptr), buf->remaining(), &bytes_read,
+                                              reinterpret_cast<uint8_t*>(_decompressed_buffer->ptr),
+                                              _decompressed_buffer->capacity, &bytes_written, &stream_end));
```
[the remark this answers is not in the document]
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.cpp:307]: Before decompressing, we have no idea about the size of decompressed data. So we have to try to decompress in pieces and put them together.

File: be/src/util/byte_buffer.h (hunk 18 above)
Comment (rickif) [reply be/src/util/byte_buffer.h:58]: Updated. It will do nothing when the new_size is less than the origin capacity.

File: be/src/runtime/stream_load/stream_load_pipe.cpp (hunk 19 above)
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.cpp]: Updated. It has been replaced with `ALIGN_UP(xxx, 1024)`

**Review by trueeyu — DISMISSED** [review]

**Review by trueeyu — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by rickif** [comment]
@mergifyio backport branch-3.3 branch-3.2

**Comment by Hatlassian** [comment]
@rickif this is really cool! Did you manage to get numbers / experiments to show how much this helped improve streaming performance?
